pipit/util/test_generator.py

```diff
@@ -0,0 +1,366 @@
+        once the entire tree has been processed.
+        """
+        data.append(
+            [
+                begin_time,
+                "Enter",
+                self.name + "()",
+                process,
+                {},
+                self.inc_time,
+                self.exc_time,
+            ]
+        )
+        if self.is_mpi:
+            data.append(
+                [
+                    begin_time + self.mpi_time,
+                    "Instant",
+                    self.mpi_type,
+                    process,
+                    self.mpi_attributes(),
+                    float("nan"),
+                    float("nan"),
+                ]
+            )
+        # total_time accumulates durations of already processed children
+        total_time = begin_time
+        for run_time, child in sorted(self.children.items()):
+            # children will add their own lines to data
+            child.to_events(total_time + run_time, process, data)
+            total_time += child.inc_time
+
+        # time.inc and time.exc are both NaN for Leave events
+        data.append(
+            [
+                begin_time + self.inc_time,
+                "Leave",
+                self.name + "()",
+                process,
+                {},
+                float("nan"),
+                float("nan"),
+            ]
+        )
+
+    def tweak_tree(self):
+        """
+        Adds small exclusive time perturbations to
+        function length to generate a "similar" tree, for other processes
+        to use, that is not identical to the original.
+        """
+        exc_time = self.exc_time
+        factor = (0.7) + random.random() * 0.6
+        exc_time_new = exc_time * factor
+        # preserve inclusive time relations
+        self.grow_inc_time(exc_time_new - exc_time)
+        self.exc_time = exc_time_new
+        for run_time, child in self.children.items():
+            child.tweak_tree()
+        # also scale back child run times
+        self.children = {
+            run_time * factor: child for run_time, child in self.children.items()
+        }
+
+    def deepcopy(self):
+        """
+        Returns a deep copy of the tree.
+        """
+        mycopy = FakeNode(
+            self.name,
+            self.exc_time,
+            self.mpi_type,
+            self.mpi_tgt,
+            self.mpi_volume,
+            self.mpi_time,
+        )
+        mycopy.inc_time = self.inc_time
+        mycopy.total_nodes = self.total_nodes
+        for run_time, child in self.children.items():
+            mycopy.children[run_time] = child.deepcopy()
+            mycopy.children[run_time].parent = mycopy
+        return mycopy
+
+    def node_at_time(self, target_time):
+        """
+        Returns the node that contains the given target_time,
+        as well as the offset into that node that the time occurs.
+        """
+        total_time = 0
+        # TODO: may be more efficient to store self.children sorted already
```

Comment:
I think this would be a good idea if we were generating very large trees, but this is not really the case for this utility. We technically could store the run times as a BST which would not require sorting, but then finding if if keys exist etc is small overhead. For now, I think this is fine.